<a href="https://colab.research.google.com/github/janiosl/C/blob/master/Exercicios/APA/apa_lista02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **CEFET-RJ**

**Programa de Pós-Gradução em Ciência da Computação - PPCIC**

**Mestrado em Ciência da Computação**

**Disciplina: Análise e Projeto de Algoritmos - 2021/2**

* **Exercícios 02**
* **Prof.: Diego Nunes Brandão**
* **Aluno: Janio de Souza Lima**

---

Uma versão online para facilitar a leitura desse notebook foi disponibilizada em: https://github.com/janiosl/C/blob/master/Exercicios/APA/apa_lista02.ipynb

## **Questão 1**:

Quantos antecedentes tem um nó no nível $n$ em uma árvore binária? Prove!

Como o nível de um nó é sempre o nível de seu pai + 1, podemos dizer que o nível a quantidade de antecedentes de um nó de nível $n$ é $n-1$, o que equivale ao nível de seu pai.

## **Questão 2**:

Escreva um programa que implementa funções para:

O código com a implementação do TAD árvore binária e as funções para atendimento às questões abaixo está disponível no anexo `apa02_02_arvBin.c` e no repositório: https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_02_arvBin.c

Abaixo de cada questão está a demonstração da execução do programa com o atendimento ao passo desejado.

a) contar o número de nós em uma árvore binária;

b) o número de folhas;

c) o número de filhos à direita;

d) a altura da árvore.


![](https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_02_arvBin_A.PNG?raw=true)

* **Código com a implementação:**

```
#include <stdio.h>
#include <stdlib.h>
/*------------------
TAD - Árvore Binária
------------------*/
struct arv{
    char info;
    struct arv* esq;
    struct arv* dir;
};typedef struct arv Arv;
/*----------------------------------------
Funções para manipulação da Árvore Binária
----------------------------------------*/
Arv* arv_criavazia(void){
    //Cria a árvore vazia
    return NULL;
}
int arv_vazia(Arv* a){
    //Verifica se a árvore é vazia
    return a==NULL;
}
Arv* arv_cria(char c, Arv* sae, Arv* sad){
    /*Cria a árvore com um nós
    e as sub-árvores sae e sad*/
    Arv* p=(Arv*)malloc(sizeof(Arv));
    p->info = c;
    p->esq = sae;
    p->dir = sad;
    return p;
}
Arv* arv_libera(Arv* a){
    //Libera a árvore da memória
    if (!arv_vazia(a)){
        arv_libera(a->esq); /* libera sae */
        arv_libera(a->dir); /* libera sad */
        free(a);            /* libera raiz */
    }
    return NULL;
}
/*----------------------------------------
Funções análise da Árvore Binária
----------------------------------------*/
void arv_imprime (Arv* a){
    //Imprime as informações da árvore
    printf("<");
    if (!arv_vazia(a)){
        printf("%c ", a->info); /* mostra raiz */
        arv_imprime(a->esq);    /* mostra sae */
        arv_imprime(a->dir);    /* mostra sad */
    }
    printf(">");
}
int arv_conta_nos(Arv* a){
    //Conta a quantidade de nós da árvore
	if(!a)
		return 0;
	else
		return 1+ arv_conta_nos(a->esq)+arv_conta_nos(a->dir);
}
int arv_num_folhas(Arv *a){
    //Conta quantidade de folhas
    int nf;
    nf = 0;
    //Verifica se a arvore é vazia
    if(!arv_vazia(a)){
        if(arv_vazia(a->esq) && arv_vazia(a->dir)){
            nf ++;
        } else{
            nf = nf + arv_num_folhas(a->esq);
            nf = nf + arv_num_folhas(a->dir);
        }
        return nf;
    }
    return 0;
}
int arv_filhos_direita(Arv *a){
    //Conta quantidade de filhos à direita
    return 0;
}
int arv_altura(Arv *a){
    //Calcula a altura da árvore
    return 0;
}
/*------------------
Programa principal
------------------*/
int main(){
    /* sub-árvore 'd' */
    Arv* a1= arv_cria('d',arv_criavazia(),arv_criavazia());
    /* sub-árvore 'b' */
    Arv* a2= arv_cria('b',arv_criavazia(),a1);
    /* sub-árvore 'e' */
    Arv* a3= arv_cria('e',arv_criavazia(),arv_criavazia());
    /* sub-árvore 'f' */
    Arv* a4= arv_cria('f',arv_criavazia(),arv_criavazia());
    /* sub-árvore 'c' */
    Arv* a5= arv_cria('c',a3,a4);
    /* árvore 'a' */
    Arv* a = arv_cria('a',a2,a5);
    /*Analisando arvore completa*/
    puts("Visualizacao da arvore a:");
    puts("-----------------------------------------");
    arv_imprime(a);
    puts("\n=========================================");
    int nos;
    int folhas;
    nos = arv_conta_nos(a);
    folhas = arv_num_folhas(a);
    printf("Nos: %d\nFolhas: %d\n\n", nos, folhas);
    /*Analisando a subárvore arvore c*/
    puts("Visualizacao da arvore c:");
    puts("-----------------------------------------");
    arv_imprime(a5);
    puts("\n=========================================");
    nos = arv_conta_nos(a5);
    folhas = arv_num_folhas(a5);
    printf("Nos: %d\nFolhas: %d\n\n", nos, folhas);
    /*Analisando uma árvore vazia*/
    Arv* vazia = arv_criavazia();
    puts("Visualizacao da arvore vazia:");
    puts("-----------------------------------------");
    nos = arv_conta_nos(vazia);
    arv_imprime(vazia);
    puts("\n=========================================");
    folhas = arv_num_folhas(vazia);
    printf("Nos: %d\nFolhas: %d\n\n", nos, folhas);
    return 0;
}
```

##3

##4

## **Questão 5**:

Observando cada um dos caracteres que formam a palavra C O M P U T A Ç Ã O. Crie uma tabela de dispersão (hash table) que armazene cada um destes caracteres, utilizando a seguinte função: $h(char) = (ordem de char no alfabeto)\%5$.

Desenhe o esquema de como ficaram a hash table e as listas encadeadas.

In [1]:
def ordem(char):
  """
  Função identifica a ordem alfabética de uma letra
  Argumentos:
      str char => Letra a ser analisada
  Retorno:
      int ord => Posição da letra no alfabeto (começando em 1)
  """
  #Tratamento de minúscula
  c = char.upper()

  #Tratamento de caracteres latinos e acentos
  if c == "Ç":
    c = 'C'
  elif c in 'ÃÁÀÂÄ':
    c = 'A'
  elif c in 'ÉÈÊË':
    c = 'E'
  elif c in 'ÍÌÎÏ':
    c = 'I'
  elif c in 'ÕÓÒÔÖ':
    c = 'O'
  elif c in 'ÚÙÛÜ':
    c = 'U'
  elif c in 'Ñ':
    c = 'N'

  #Função hash
  #Define a ordem da letra no alfabeto
  alfa = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  ord = alfa.find(c)+1

  return ord


def h(char, ord):
  """
  Função Calcula a função hash para definir a posição da letra na tabela
  Argumentos:
      str char => Letra a ser analisada
      int ord => Ordem da letra no alfabeto
  Retorno:
      dict hash => Dicionário contendo:
          str char => Letra analisada
          int pos => Calculada baseada na função h(char) = (ordem alfabética)%5
  """
  pos = int(ord % 5)
  hash = {pos: char}

  return hash

In [2]:
#Teste da aplicação da função coma diferentes grafias da mesma palavra
palavra = 'COMPUTAÇÃO'
#palavra = 'COmPUTAÇÃO'
#palavra = 'computação'

In [3]:
#Aplicação da função hash e exibição da posição de cada letra
print('Caractere : Posição')
for letra in palavra:
  o = ordem(letra)
  hs = h(letra, o)
  for k, v in hs.items():
    print(f'{v:^10}:{k:^8}')

Caractere : Posição
    C     :   3    
    O     :   0    
    M     :   3    
    P     :   1    
    U     :   1    
    T     :   0    
    A     :   1    
    Ç     :   3    
    Ã     :   1    
    O     :   0    


Percebe-se que com a função hash definida por $h(char) =$ (ordem alfabética)$\%5$ a tabela construída conterá diversas colisões. Por exemplo, a posição $3$ conterá os caracteres C, M e Ç. Podemos ver em mais detalhes a distribuição dos caracteres na tabela a seguir, que evidencia ainda que poucas posições foram usadas para alocação dos elementos na *hash table*.

| Posição  |  Caractere  |
| -------- | ----------- |
|  0 |  O, T, O |
|  1 |  P, U, A, Ã |
|  3 |  C, M, Ç |

Seria interessante a aplicação de uma técnica de tratamento de colisção para evitar esta situação. Por exemplo, posicionando o elemento (caractere) na próxima posição livre quando o resultado da funçao *hash* aponta para uma posição ocupada, ou seja, à medida que colisões fossem identificadas. Esta abordagem é conhecida como **endereçamento aberto** e uma aplicação manual dela em nosso exemplo, resultaria na tabela abaixo:

|Hash | Posição  |  Caractere  |
|---| ---| ----------- |
| 0 |  0 |  O |
| 1 |  1 |  P |
| 1 |  2 |  U |
| 3 |  3 |  C |
| 3 |  4 |  M |
| 0 |  5 |  T |
| 1 |  6 |  A |
| 3 |  7 |  Ç |
| 1 |  8 |  Ã |
| 0 |  9 |  O |

A seguir apresento também uma implementação em `C` de uma versão simplificada da função de hash e sua aplicação.

```
#include <stdio.h>
#include <stdlib.h>
/*--------------------------------------------------------------------------
Função Calcula a função hash para definir a posição da letra na tabela
Argumentos:
    int ord => Ordem da letra no alfabeto
Retorno:
    int pos => Calculada baseada na função h(char) = (ordem alfabética)%5
--------------------------------------------------------------------------*/
int h(int ord){
    int pos;
    pos = ord % 5;
    return pos;
}
/*--------------------------------------------------------------------------
Programa Principal
--------------------------------------------------------------------------*/
int main(){
    char letra;
    puts("Letra: ");
    scanf("%c", &letra);
    int ord;
    ord = ordem(letra);
    int pos;
    pos = h(ord);
    puts("----------------------");
    puts("Letra : Pos Hash Table");
    printf("  %c   :       ", letra);
    printf("%d\n", pos);
    puts("----------------------");
    return 0;
}
```


**OBSERVAÇÃO:** Na implementação abaixo não está implementada a função `int ordem(char letra)` que ainda está em desenvolvimento. Mas a lógica é a mesma da função implementada em Python acima (`ordem(char)`), ou seja, conferir a ordem alfabética da letra desejada e passar esta posição como argumento para função `h(ord)` que retornará um valor inteiro com a posição definida pela função *hash*.

O resultado da execução em C está demonstrado na imagem a seguir, realizada usando o código acima com a função de ordenação parcial/incompleta (https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_05_hashComp.c), mas que já demonstra a parte principal que é a aplicação da função *hash*.

![](https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_05_hashComp.PNG?raw=true)

## **Questão 6**:

Crie um arquivo hash com registros para $25$ nomes de cidades do estado de Minas Gerais cujos nomes comecem com as seguintes letras do alfabeto: a, b, c e s. A chave de cada registro é o nome da cidade, não sendo necessários outros campos. Coloque inicialmente os nomes das cidades em ordem alfabética.

###6.1 Obtenção dos dados

Inicialmente os trechos abaixo realizam apenas a varredura de uma página na internet (Wikipedia) com uma tabela com a lista de todos os municípios de Minas Gerais. Esta etapa será apenas para carga dos dados que serão usados nas demais etapas da questão.

In [33]:
#Carga dos dados das cidades
#Página wikipedia com cidades de Minas Gerais
url = 'https://pt.wikipedia.org/wiki/Lista_de_munic%C3%ADpios_de_Minas_Gerais_por_popula%C3%A7%C3%A3o'

In [34]:
#Bibliotecas para carga de dados via web scraping
from bs4 import BeautifulSoup
import requests
#Biblioteca para limpeza do texto
import re

In [35]:
def cidades_minas(url, q, iniciais):
  #Carga da tabela de ciades da url desejada
  page = requests.get(url)
  bs = BeautifulSoup(page.content, 'html.parser')
  tabela = bs.find_all('td')

  #Captura do texto da tabela
  cidades = []
  for k in range(0, len(tabela)):
    cidades.append(tabela[k].get_text())

  #Limpeza do texto (eliminação de quebra de linhas)
  cidades = [re.sub('\n', '', cidades[i]) for i in range(0, len(cidades))]

  #Eliminação das linhas indesejadas (posição e população)
  #Seleção cidades com iniciais desejadas
  c = []
  for k in range(0, len(cidades)):
    #Selecionada linhas iniciadas pelas letras desejadas
    if cidades[k][0].lower() in iniciais.lower():
      c.append(cidades[k])

  #Seleciona apenas as q primeiras cidades
  cidades = c[:q]

  return cidades

In [36]:
q = 25
letras = 'abcs'
cidades = cidades_minas(url, q, letras)

In [37]:
#Visualização da lista de cidades
cidades.sort()
cidades

['Alfenas',
 'Araguari',
 'Araxá',
 'Barbacena',
 'Belo Horizonte',
 'Betim',
 'Bocaiúva',
 'Bom Despacho',
 'Caeté',
 'Campo Belo',
 'Caratinga',
 'Cataguases',
 'Congonhas',
 'Conselheiro Lafaiete',
 'Contagem',
 'Coronel Fabriciano',
 'Curvelo',
 'Sabará',
 'Santa Luzia',
 'Santos Dumont',
 'Sete Lagoas',
 'São Francisco',
 'São João del Rei',
 'São Lourenço',
 'São Sebastião do Paraíso']

Analisando de maneira superficial a lista ordenada acima, notamos que algumas situações podem influenciar a predisposição para colisões no uso de funções *hash* baseadas nas letras dos nomes das cidades, conforme segue:
* Usar apenas a letra inicial, obviamente, gerará diversas colisões, tendo em vista que a lista resume-se a cidades iniciadas apenas com quatro letras (a, b, c e c).
* Mesmo aumentando a quantidade de caracters teremos muitas situações de repetiação nas primeiras quantidade, por exemplo para 3 caracteres teremos diversas cidades iniciadas por "São", "San", "Ara" e "Con".
* A situação acima agrava-se com o uso de 2 caractesres, pois além dos exemplos já citados, acrescenta-se alguns nomes iniciados por "Be", "Sa".

Nas próximas seções adotaremos duas opções de função *hash* e avaliaremos na prática se os indícios acima se confirma.

###6.2 Função Hash: Experimentos

Primeiro experimento:

$h(cidade) \approx \sqrt{\sum_{i=1}^n{unicode(Letra_i(cidade))}}$

A função *hash* proposta será calculada pela raiz quadrada do somatório das $i$-ésimas letras. Sendo que $i$ vai de $1$ até a quantidade $n$,  definida no momento da aplicação *hash*, $n \leq$ tamanho do nome cidade. Usamos o símbolo de aproximado ($\approx$) ao invés de igual ($=$), pois na função estamos arredondando o resultado para um número inteiro (com a função interna do `int`), tendo em vista que o resultado deverá ser uma posição exata em nossa *hash table*.

In [38]:
def h_cidades(nome, n):
  """
  Função hash baseada na raiz quadrada do somatório do unicode das n letras
  """
  #Checa o tamanho de n
  if n > len(nome):
    print(f'Tamanho de N ({n}) inválido para cidade {nome}')
    return None

  #Aplica a função hash definida
  h = 0
  for k in range(0, n):
    h += ord(nome[k])

  h = int(h**(1/2))

  return h

In [39]:
h_cidades(cidades[0], 5)

22

In [40]:
def gera_hash_table(cidades, N, imprimir):
  #Hash table vazia para receber os dados
  hash_table_cidades = dict()
  #Variáveis de controle das iterações
  hTemp = []

  for c in cidades:
    h = h_cidades(c, N)
    if h not in hTemp:
      hash_table_cidades[h] = [c]
      hTemp.append(h)
    else:
      for k, v in hash_table_cidades.items():
        if k==h:
          v.append(c)

    #Imprime passo a passo na tela se o parâmetro estiver configurado = True
    if imprimir:
      print(f'{h} : {c}')
  
  return hash_table_cidades

Experimento 1.1 => $N=5$

In [41]:
N = 5
ht = gera_hash_table(cidades, N, True)

22 : Alfenas
22 : Araguari
24 : Araxá
21 : Barbacena
20 : Belo Horizonte
22 : Betim
21 : Bocaiúva
19 : Bom Despacho
24 : Caeté
22 : Campo Belo
22 : Caratinga
21 : Cataguases
22 : Congonhas
22 : Conselheiro Lafaiete
22 : Contagem
22 : Coronel Fabriciano
22 : Curvelo
22 : Sabará
22 : Santa Luzia
22 : Santos Dumont
20 : Sete Lagoas
22 : São Francisco
22 : São João del Rei
23 : São Lourenço
23 : São Sebastião do Paraíso


In [42]:
#Conferência do resultado
ht

{19: ['Bom Despacho'],
 20: ['Belo Horizonte', 'Sete Lagoas'],
 21: ['Barbacena', 'Bocaiúva', 'Cataguases'],
 22: ['Alfenas',
  'Araguari',
  'Betim',
  'Campo Belo',
  'Caratinga',
  'Congonhas',
  'Conselheiro Lafaiete',
  'Contagem',
  'Coronel Fabriciano',
  'Curvelo',
  'Sabará',
  'Santa Luzia',
  'Santos Dumont',
  'São Francisco',
  'São João del Rei'],
 23: ['São Lourenço', 'São Sebastião do Paraíso'],
 24: ['Araxá', 'Caeté']}

In [43]:
#Visualização detalhada para análise de colisões
print('Posição : Colisões : Elementos  ')
for k, v in ht.items():
  if k != None:
    print(f'{k:^8}:{len(v)-1:^10}: {v}')
  else:
    print(f'  {k}  :{len(v)-1:^10}: {v}')

Posição : Colisões : Elementos  
   22   :    14    : ['Alfenas', 'Araguari', 'Betim', 'Campo Belo', 'Caratinga', 'Congonhas', 'Conselheiro Lafaiete', 'Contagem', 'Coronel Fabriciano', 'Curvelo', 'Sabará', 'Santa Luzia', 'Santos Dumont', 'São Francisco', 'São João del Rei']
   24   :    1     : ['Araxá', 'Caeté']
   21   :    2     : ['Barbacena', 'Bocaiúva', 'Cataguases']
   20   :    1     : ['Belo Horizonte', 'Sete Lagoas']
   19   :    0     : ['Bom Despacho']
   23   :    1     : ['São Lourenço', 'São Sebastião do Paraíso']


Experimento 1.2 => $N=3$

In [44]:
N = 3
ht = gera_hash_table(cidades, N, False)

In [45]:
#Visualização detalhada para análise de colisões
print('Posição : Colisões : Elementos  ')
for k, v in ht.items():
  if k != None:
    print(f'{k:^8}:{len(v)-1:^10}: {v}')
  else:
    print(f'  {k}  :{len(v)-1:^10}: {v}')

Posição : Colisões : Elementos  
   16   :    15    : ['Alfenas', 'Araguari', 'Araxá', 'Barbacena', 'Belo Horizonte', 'Betim', 'Bocaiúva', 'Bom Despacho', 'Caeté', 'Campo Belo', 'Caratinga', 'Cataguases', 'Congonhas', 'Conselheiro Lafaiete', 'Contagem', 'Sabará']
   17   :    4     : ['Coronel Fabriciano', 'Curvelo', 'Santa Luzia', 'Santos Dumont', 'Sete Lagoas']
   20   :    3     : ['São Francisco', 'São João del Rei', 'São Lourenço', 'São Sebastião do Paraíso']


Experimento 1.3 => $N=7$

In [46]:
N = 7
ht = gera_hash_table(cidades, N, True)

26 : Alfenas
26 : Araguari
Tamanho de N (7) inválido para cidade Araxá
None : Araxá
25 : Barbacena
24 : Belo Horizonte
Tamanho de N (7) inválido para cidade Betim
None : Betim
29 : Bocaiúva
24 : Bom Despacho
Tamanho de N (7) inválido para cidade Caeté
None : Caeté
24 : Campo Belo
26 : Caratinga
26 : Cataguases
26 : Congonhas
26 : Conselheiro Lafaiete
26 : Contagem
26 : Coronel Fabriciano
27 : Curvelo
Tamanho de N (7) inválido para cidade Sabará
None : Sabará
24 : Santa Luzia
25 : Santos Dumont
24 : Sete Lagoas
27 : São Francisco
29 : São João del Rei
27 : São Lourenço
27 : São Sebastião do Paraíso


In [47]:
#Visualização detalhada para análise de colisões
print('Posição : Colisões : Elementos  ')
for k, v in ht.items():
  if k != None:
    print(f'{k:^8}:{len(v)-1:^10}: {v}')
  else:
    print(f'  {k}  :{len(v)-1:^10}: {v}')

Posição : Colisões : Elementos  
   26   :    7     : ['Alfenas', 'Araguari', 'Caratinga', 'Cataguases', 'Congonhas', 'Conselheiro Lafaiete', 'Contagem', 'Coronel Fabriciano']
  None  :    3     : ['Araxá', 'Betim', 'Caeté', 'Sabará']
   25   :    1     : ['Barbacena', 'Santos Dumont']
   24   :    4     : ['Belo Horizonte', 'Bom Despacho', 'Campo Belo', 'Santa Luzia', 'Sete Lagoas']
   29   :    1     : ['Bocaiúva', 'São João del Rei']
   27   :    3     : ['Curvelo', 'São Francisco', 'São Lourenço', 'São Sebastião do Paraíso']


**Análise dos resultados:**

Com a função *hash* usada neste primeiro experimento, percebemos que há um excessivo número de colisões, mesmo com o uso de diferentes valores para $n$. Além disso, quando eleva-se muito o valor de $n$, as colisões tenderam a diminuir, mas algumas cidades ficaram sem retorno válido, tendo em vista que o valor de $n$ excedeu a quantidade de caracteres do nome da cidade.


Segundo experimento:

$h(cidade) \approx \frac{\sum_{i=1}^n{unicode(Letra_i(cidade))}}{n} $

A função *hash* proposta será calculada pela raiz quadrada do somatório das $i$-ésimas letras. Sendo que $i$ vai de $1$ até a quantidade $n$,  definida no momento da aplicação *hash*, $n \leq$ tamanho do nome cidade. Usamos o símbolo de aproximado ($\approx$) ao invés de igual ($=$), pois na função estamos arredondando o resultado para um número inteiro (com a função interna do `int`), tendo em vista que o resultado deverá ser uma posição exata em nossa *hash table*.

In [48]:
def h_cidades(nome, n):
  """
  Função hash baseada na divisão inteira
  do somatório do unicode das n letras por n
  """
  #Checa o tamanho de n
  if n > len(nome):
    print(f'Tamanho de N ({n}) inválido para cidade {nome}')
    return None

  #Aplica a função hash definida
  h = 0
  for k in range(0, n):
    h += ord(nome[k])

  h = int(h/n)

  return h

Experimento 2.1 => $N=5$

In [49]:
N = 5
ht = gera_hash_table(cidades, N, True)

97 : Alfenas
99 : Araguari
124 : Araxá
94 : Barbacena
83 : Belo Horizonte
99 : Betim
95 : Bocaiúva
77 : Bom Despacho
122 : Caeté
99 : Campo Belo
98 : Caratinga
96 : Cataguases
100 : Congonhas
100 : Conselheiro Lafaiete
100 : Contagem
102 : Coronel Fabriciano
103 : Curvelo
97 : Sabará
100 : Santa Luzia
103 : Santos Dumont
86 : Sete Lagoas
104 : São Francisco
105 : São João del Rei
105 : São Lourenço
107 : São Sebastião do Paraíso


In [50]:
#Visualização detalhada para análise de colisões
print('Posição : Colisões : Elementos  ')
for k, v in ht.items():
  if k != None:
    print(f'{k:^8}:{len(v)-1:^10}: {v}')
  else:
    print(f'  {k}  :{len(v)-1:^10}: {v}')

Posição : Colisões : Elementos  
   97   :    1     : ['Alfenas', 'Sabará']
   99   :    2     : ['Araguari', 'Betim', 'Campo Belo']
  124   :    0     : ['Araxá']
   94   :    0     : ['Barbacena']
   83   :    0     : ['Belo Horizonte']
   95   :    0     : ['Bocaiúva']
   77   :    0     : ['Bom Despacho']
  122   :    0     : ['Caeté']
   98   :    0     : ['Caratinga']
   96   :    0     : ['Cataguases']
  100   :    3     : ['Congonhas', 'Conselheiro Lafaiete', 'Contagem', 'Santa Luzia']
  102   :    0     : ['Coronel Fabriciano']
  103   :    1     : ['Curvelo', 'Santos Dumont']
   86   :    0     : ['Sete Lagoas']
  104   :    0     : ['São Francisco']
  105   :    1     : ['São João del Rei', 'São Lourenço']
  107   :    0     : ['São Sebastião do Paraíso']


Experimento 2.2 => $N=3$

In [51]:
N = 3
ht = gera_hash_table(cidades, N, True)

91 : Alfenas
92 : Araguari
92 : Araxá
92 : Barbacena
91 : Belo Horizonte
94 : Betim
92 : Bocaiúva
95 : Bom Despacho
88 : Caeté
91 : Campo Belo
92 : Caratinga
93 : Cataguases
96 : Congonhas
96 : Conselheiro Lafaiete
96 : Contagem
97 : Coronel Fabriciano
99 : Curvelo
92 : Sabará
96 : Santa Luzia
96 : Santos Dumont
100 : Sete Lagoas
140 : São Francisco
140 : São João del Rei
140 : São Lourenço
140 : São Sebastião do Paraíso


In [52]:
#Visualização detalhada para análise de colisões
print('Posição : Colisões : Elementos  ')
for k, v in ht.items():
  if k != None:
    print(f'{k:^8}:{len(v)-1:^10}: {v}')
  else:
    print(f'  {k}  :{len(v)-1:^10}: {v}')

Posição : Colisões : Elementos  
   91   :    2     : ['Alfenas', 'Belo Horizonte', 'Campo Belo']
   92   :    5     : ['Araguari', 'Araxá', 'Barbacena', 'Bocaiúva', 'Caratinga', 'Sabará']
   94   :    0     : ['Betim']
   95   :    0     : ['Bom Despacho']
   88   :    0     : ['Caeté']
   93   :    0     : ['Cataguases']
   96   :    4     : ['Congonhas', 'Conselheiro Lafaiete', 'Contagem', 'Santa Luzia', 'Santos Dumont']
   97   :    0     : ['Coronel Fabriciano']
   99   :    0     : ['Curvelo']
  100   :    0     : ['Sete Lagoas']
  140   :    3     : ['São Francisco', 'São João del Rei', 'São Lourenço', 'São Sebastião do Paraíso']


Experimento 2.3 => $N=7$

In [53]:
N = 7
ht = gera_hash_table(cidades, N, True)

99 : Alfenas
101 : Araguari
Tamanho de N (7) inválido para cidade Araxá
None : Araxá
96 : Barbacena
85 : Belo Horizonte
Tamanho de N (7) inválido para cidade Betim
None : Betim
120 : Bocaiúva
86 : Bom Despacho
Tamanho de N (7) inválido para cidade Caeté
None : Caeté
84 : Campo Belo
100 : Caratinga
99 : Cataguases
102 : Congonhas
102 : Conselheiro Lafaiete
100 : Contagem
103 : Coronel Fabriciano
105 : Curvelo
Tamanho de N (7) inválido para cidade Sabará
None : Sabará
87 : Santa Luzia
94 : Santos Dumont
86 : Sete Lagoas
104 : São Francisco
123 : São João del Rei
108 : São Lourenço
105 : São Sebastião do Paraíso


In [54]:
#Visualização detalhada para análise de colisões
print('Posição : Colisões : Elementos  ')
for k, v in ht.items():
  if k != None:
    print(f'{k:^8}:{len(v)-1:^10}: {v}')
  else:
    print(f'  {k}  :{len(v)-1:^10}: {v}')

Posição : Colisões : Elementos  
   99   :    1     : ['Alfenas', 'Cataguases']
  101   :    0     : ['Araguari']
  None  :    3     : ['Araxá', 'Betim', 'Caeté', 'Sabará']
   96   :    0     : ['Barbacena']
   85   :    0     : ['Belo Horizonte']
  120   :    0     : ['Bocaiúva']
   86   :    1     : ['Bom Despacho', 'Sete Lagoas']
   84   :    0     : ['Campo Belo']
  100   :    1     : ['Caratinga', 'Contagem']
  102   :    1     : ['Congonhas', 'Conselheiro Lafaiete']
  103   :    0     : ['Coronel Fabriciano']
  105   :    1     : ['Curvelo', 'São Sebastião do Paraíso']
   87   :    0     : ['Santa Luzia']
   94   :    0     : ['Santos Dumont']
  104   :    0     : ['São Francisco']
  123   :    0     : ['São João del Rei']
  108   :    0     : ['São Lourenço']


**Análise dos resultados:**

Com a função *hash* usada no segundo experimento, foi possível notar uma redução das colisões. Mesmo com o menor valor de $n$ escolhido ($3$), houve muitas posições sem colisão. Com os valores de $n$ maiores ($5, 7$), os elementos ficaram melhor distribuídos, o que facilitaria por exemplo o uso de técnicas de resolução de colisões sem excessiva diferença entre as posições calculadas inicialmente e as alteradas pela técnica escolhida. Por fim, mantém-se a atenção quando eleva-se muito o valor de $n$, pois algumas cidades ficam sem retorno válido, tendo em vista que o valor de $n$ excedeu a quantidade de caracteres do nome da cidade.

Exemplo de implementações das duas implementações abaixo das funções *hash* em C estão no anexo `apa02_06_hashCid.c` e no repositório: https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_06_hashCid.c.

```
#include <stdio.h>
float raiz(float numero){
    float precisao = 0.000001;
    float b = numero, a=1;
    while ((b-a) >= precisao){
        b = (a+b)/2;
        a= numero / b;
    }
    return b;
}
int h_cidades_raiz(char *nome, int n){
    int asc;
    int i;
    int h = 0;
    //Cálculo do somatório do código ASCII
    for(i=0; i<n; i++){
        asc = nome[i];
        h = h + asc;
    }
    //Cálculo da raiz quadrada do somatório
    return raiz(h);
}
int h_cidades_div(char *nome, int n){
    int asc;
    int i;
    int h = 0;
    //Cálculo do somatório do código ASCII
    for(i=0; i<n; i++){
        asc = nome[i];
        h = h + asc;
    }
    //Cálculo do somatório dos códigos ASCII / n
    return h/n;
}
int main(){
    char cidades[25][100] = {
        "Alfenas","Araguari","Araxá","Barbacena","Belo Horizonte",
        "Betim","Bocaiúva","Bom Despacho","Caeté","Campo Belo",
        "Caratinga","Cataguases","Congonhas","Conselheiro Lafaiete",
        "Contagem","Coronel Fabriciano","Curvelo","Sabará","Santa Luzia",
        "Santos Dumont","Sete Lagoas","São Francisco","São João del Rei",
        "São Lourenço","São Sebastião do Paraíso"};
    char *nome;
    nome = cidades[0];
    int pos;
    int n = 3;
    puts("==================================");
    puts("Teste simples");
    puts("==================================");
    //Aplicação simples da primeira função hash
    printf("Hash (op. 1) para N = %d\n", n);
    puts("--------------------------");
    pos = h_cidades_raiz(nome, n);
    printf("%d : \"%s\"\n", pos, nome);
    //Aplicação simples da segunda função hash
    printf("Hash (op. 2) para N = %d\n", n);
    puts("--------------------------");
    pos = h_cidades_div(nome, n);
    printf("%d : \"%s\"\n", pos, nome);
    //Aplicação da segunda função hash para todas as cidades
    puts("==================================");
    puts("Aplicacao para todas as cidades");
    puts("==================================");
    int j;
    printf("Hash (op. 2) para N = %d\n", n);
    puts("--------------------------");
    for(j=0; j<25; j++){
        pos = h_cidades_div(cidades[j], n);
        printf("%d : \"%s\"\n", pos, cidades[j]);
    }
    return 0;
}
```

Conforme imagem demonstrando o resultado de sua execução, a princípio o resultado foi bem sucedido, contudo, nas palavras com caracteres acentuados há uma divergência, talvez pela por alguma interpretação diferente entre o compilador usado e o interpretador do Python.

![](https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_06_hashCid.PNG?raw=true)

Conferindo os resultados manualmente para $n=3$ e alguns nomes de cidades iniciadas por "São", temos o seguinte:

|Caractere|ASCII|Função ord()$^1$|
|---------|-----|----------------|
|S|83|83|
|ã|-58|227|
|o|111|111|
|TOTAL|136|421|
|*$^1$ Função usada no Python para cálculo do código de cada caractere.*|

* `C`: Aplicação do complemento da função de hash 1: $\sqrt{136} \approx 12$
* `Python`: Aplicação do complemento da função de hash 1: $\sqrt{421} \approx 20$
* `C`: Aplicação do complemento da função de hash 2: $\frac{136}{n} = \frac{136}{3} \approx 45$
* `Python`: Aplicação do complemento da função de hash 2: $\frac{421}{n} = \frac{421}{3} \approx 140$

Percebe-se que os resultados para estes nomes estão coerentes com o resultado a função `ord()` em Python, apesar de serem diferentes dos códigos ASCII obtidos em `C`. Nesta última lingaugem, por usa vez, os cálculos manuais não estão conferindo com o cálculo pela função implementada, isto pode estar ocorrendo por conta do caractere acentuado (ã) não ter sido interpretado da mesma foram ao digitar manualmente e ao ler do vetor de caracteres. De qualquer modo, a lógica geral já apresentada nas partes anteriores da questão já demonstram o uso da função, mas consideri relevante a adição deste trecho para evidenciar o cuidado no uso das funções *hash* para garantir que a interpretação dos elementos serão usados para seu cálculo serão interpretados adequadamente, bem como nos cuidados na adaptação entre diferentes plataformas ou linguagens, que podem levar em um caso real a conclusões indevidas ou dificuldades de resolução/recuperação da informação original tratada com um *hash*.

Demonsração da cálculo manual dos códigos ASCII usando script complementar (listado abaixo) para identificação do código de determinando caractere.

![](https://github.com/janiosl/C/blob/master/Exercicios/APA/apa02_06_hashCid-conf.PNG?raw=true)

```
#include <stdio.h>
int main(){
    int asc;
    char caractere;
    puts("Identificar codigo ASCII");
    puts("=====================================");
    printf("Digite o caractere desejado: ");
    scanf("%c", &caractere);
    //Identificação do código ascii do caractere desejado
    asc = caractere;
    //Exibição do resultado
    printf("Codigo ASCII para caractere \"%c\": %d\n", caractere, asc);
    puts("=====================================");
    return 0;
}
```



##7